compiler/rustc_hir_analysis/src/coherence/unsafety.rs

```diff
@@ -20,7 +20,20 @@ pub(super) fn check_item(
         tcx.generics_of(def_id).own_params.iter().find(|p| p.pure_wrt_drop).map(|_| "may_dangle");
     let trait_ref = trait_header.trait_ref.instantiate_identity();
 
-    match (trait_def.safety, unsafe_attr, trait_header.safety, trait_header.polarity) {
+    let is_copy = tcx.is_lang_item(trait_def.def_id, LangItem::Copy);
+    let trait_def_safety = if is_copy {
+        use rustc_type_ir::inherent::*;
```

Comment:
don't do this lol

Reply:
inherent is only for the solver

Reply:
What's your recommendation? The intent of this PR is to implement the suggestion in the RFC discussion that `Copy` is conditionally (un)safe to implement depending on whether the `Self` type has unsafe fields. Is there a better way to achieve this?

Reply:
Ah gotcha. Hm.

Reply:
just duplicate the method on ty or whatever

Reply:
I import `inherent::Ty` here because it provides the `has_unsafe_fields` helper. Should I also add that function as an inherent method on `rustc_middle::ty::Ty`?

Reply:
Will do!

Reply:
[Done](https://github.com/rust-lang/rust/compare/96fc9a9c59cdd41e6c003ff6409b176cc2264c8b..47454f3d33372ab04fdcc83718cae979785d4663)